enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -54,37 +54,24 @@ pub struct Handler<'a> {
 
 impl<'a> Write for Handler<'a> {
     fn write_str(&mut self, s: &str) -> core::fmt::Result {
-        let bytes = s.as_bytes();
-        let len = bytes.len() as _;
-        if len == 0 {
+        if s.as_bytes().len() == 0 {
             return Ok(());
         }
 
-        // Allocate some unencrypted memory.
-        let map = match self.ualloc(len) {
-            Err(_) => return Err(core::fmt::Error),
-            Ok(map) => map,
+        let mut cursor = self.block.cursor();
+        let untrusted = match cursor.copy_slice(s.as_bytes()).or(Err(core::fmt::Error)) {
+            Ok(slice) => slice,
+            Err(e) => return Err(e),
         };
 
-        unsafe {
-            // Copy the encrypted input into unencrypted memory.
-            core::ptr::copy_nonoverlapping(bytes.as_ptr(), map, bytes.len());
-
-            // Do the syscall; replace encrypted memory with unencrypted memory.
-            self.syscall(
-                libc::SYS_write as u64,
-                libc::STDERR_FILENO as u64,
-                map as _,
-                len,
-                0,
-                0,
-                0,
-            );
+        // Do the syscall; replace encrypted memory with unencrypted memory.
+        let req = request!(libc::SYS_write => libc::STDERR_FILENO, untrusted, s.as_bytes().len());
+        let res = unsafe { self.proxy(req) };
 
-            self.ufree(map, len);
+        match res {
+            Ok(_) => Ok(()),
+            Err(_) => Err(core::fmt::Error),
```

Comment:
`write` might return a number of bytes that is less than the number you set out to write. Does the shim care in this case? I know it's just a shim-print routine.

Reply:
Good point, I've included a check for this. It returns an error if the two aren't equal, because I figure if it writes more bytes than we expect then that's also bad. I wasn't able to use `untrusted.len()` there because of the borrow checker.